docs: Spawn direct cross-script calling (#494)
nervosnetwork/docs.nervos.org: website/docs/script/spawn_direct_cross-script_calling.mdx

---
id: spawn-cross-script-calling
title: "Spawn: Direct Cross-Script Calling"
---

import Tooltip from "@components/Tooltip";

# Spawn: Direct Cross-Script Calling

> The Meepo hardfork has introduced a range of enhancements into CKB Script development, with one major innovation being Spawn.
> 
> In [CKB Script](https://docs.nervos.org/docs/script/intro-to-script), Spawn refers to a concept and a set of syscalls for creating new processes. Unlike the [exec syscall](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec) introduced during [CKB’s first hard fork in 2022](https://archive.nervos.org/blog/nervos-layer-1-a-major-protocol-upgrade-is-rolling-out), Spawn enables **direct cross-script calls**, simplifying development process while offering greater control and optimization.

Reviewer: ```suggestion
The Meepo hard fork in 2024 has introduced a range of enhancements into CKB Script development, with one major innovation being Spawn.

In [CKB Script](/docs/script/intro-to-script), Spawn refers to a concept and a set of syscalls for creating new processes. Unlike the [exec syscall](https://github.com/nervosnetwork/rfcs/blob/master/rfcs/0034-vm-syscalls-2/0034-vm-syscalls-2.md#exec) introduced during [CKB’s first hard fork in 2021](/docs/history-and-hard-forks/ckb-hard-fork-history#1st-hard-fork--ckb-edition-mirana-2021), Spawn enables **direct cross-script calls**, simplifying development process while offering greater control and optimization.
```

## Why Spawn

> The exec syscall, introduced into <Tooltip>CKB-VM</Tooltip> during [CKB’s first hard fork](https://archive.nervos.org/blog/nervos-layer-1-a-major-protocol-upgrade-is-rolling-out), was designed to enable the loading and executing of a new Script within the current execution space. While powerful, exec can be inefficient, especially when preserving the current execution space is necessary. With exec, all state information of the current Script is discarded. When script A calls Script B using exec, Script A's context is lost, as illustrated below:
> 

Reviewer: ```suggestion
The exec syscall, introduced into <Tooltip>CKB-VM</Tooltip> during [CKB’s first hard fork](/docs/history-and-hard-forks/ckb-hard-fork-history#1st-hard-fork--ckb-edition-mirana-2021), was designed to enable the loading and executing of a new Script within the current execution space. While powerful, exec can be inefficient, especially when preserving the current execution space is necessary. With exec, all state information of the current Script is discarded. When Script A calls Script B using exec, Script A's context is lost, as illustrated below:

```

```
      +----------+   Exec   +----------+
----->+ Script A +--------->+ Script B +---->
      +----------+          +----------+
```

We aim to preserve the context of Script A. Ideally, when Script B executes, Script A should be able to retrieve B's execution result and then continue its own execution. This process is illustrated below:

```
      +----------+
----->+ Script A +--------------+----------->
      +-----+----+              ^
            |                   |
            | Spawn             | Exit
            v                   |
      +----------+              |
      | Script B +--------------+
      +----------+
```

This is where Spawn comes in, fulfilling exactly this purpose.

## Development of Spawn in CKB Virtual Machine

The Spawn syscalls in the CKB-VM draw inspiration from Linux standards, offering an efficient method for process creation and using familiar terms like "process," "pipe," and "file descriptor." We've dedicated considerable effort to make this new technology user-friendly, enabling developers to build multi-process applications with the same ease as on a local system.

> :::note
> 
> Please note that what "process" refers to in CKB-VM is different from that in an operating system: in CKB-VM, each running Script is considered a process, essentially an entity with its own independent execution space. However, at the OS level, CKB-VM remains a single-process, single-threaded program.
> 
> :::

Reviewer: ```suggestion
:::note

The term "process" in CKB-VM differs from its meaning in an operating system: in CKB-VM, each running Script is considered a process, essentially an entity with its own independent execution space. However, at the OS level, CKB-VM remains a single-process, single-threaded program.

:::
```

  - 2: output Cells
  - 0x0100000000000002: output Cells with the same running Script as current Script
  - 3: dep Cells
- **place**: A value of 0 or 1:
  - 0: read from Cell data
  - 1: read from witness
- **bounds**: high 32 bits means offset; low 32 bits means length. If length equals to zero, it reads to end instead of reading 0 byte.
- **spawn_args**: pass data during process creation or save return data. It is a structure with four members:
  - **argc**: contains the number of arguments passed to the program
  - **argv**: a one-dimensional array of strings
  - **process_id**: a pointer used to save the process_id of the child process
  - **inherited_fds**: an array representing the file descriptors passed to the child process. It must end with zero. For example, if you want to pass fd1 and fd2, you need to construct an array [fd1, fd2, 0].

The arguments used here—index, source, bounds, place, argc, and argv—are identical to those used in `exec()`.

### `pipe()`

Like Linux, CKB-VM uses pipes and file descriptors to pass data between processes. Here's how it works:

```rust
pub fn pipe() -> Result<(u64, u64), SysError>;
```

> The pipe() function creates a pipe—a unidirectional data channel used for interprocess communication. It returns a tuple of two file descriptors representing the pipe's ends. The first descriptor refers to the read end, while the second refers to the write end. When data is written to the write end, the CKB-VM buffers it until it's read from the read end.

Reviewer: ```suggestion
The `pipe()` function creates a pipe—a unidirectional data channel used for interprocess communication. It returns a tuple of two file descriptors representing the pipe's ends. The first descriptor refers to the read end, while the second refers to the write end. Data written to the write end is buffered by CKB-VM until it is read from the read end.
```

```rust
pub fn inherited_fds() -> Vec<u64>;
```

### `close`

This syscall manually closes a file descriptor. After calling it, any attempts to read from or write to the file descriptor at the other end will fail.

```
pub fn close(fd: u64) -> Result<(), SysError>;
```

### `wait()`

The last function is `wait()`. The syscall pauses execution until the process specified by `pid` has completed.

```rust
pub fn wait(pid: u64) -> Result<i8, SysError>;
```

## Errors

The Spawn family of syscalls can return five distinct errors. Most of them are self-explanatory and typically occur when you pass incorrect arguments or attempt to double-close a file descriptor.

```rust
pub enum SysError {
    /// Failed to wait. Its value is 5.
    WaitFailure,
    /// Invalid file descriptor. Its value is 6.
    InvalidFd,
    /// Reading from or writing to file descriptor failed due to other end closed. Its value is 7.
    OtherEndClosed,
    /// Max vms has been spawned. Its value is 8.
    MaxVmsSpawned,
    /// Max fds has been created. Its value is 9.
    MaxFdsCreated,
}
```

> Among these errors, a notable one is OtherEndClosed. It allows us to implement a function like `read_all(fd) -> Vec<u8>` to read all the data from the pipe at once. Many programming languages provide similar functions in their standard libraries, and you can easily implement this functionality yourself in the CKB-VM.

Reviewer: ```suggestion
Among these errors, a notable one is `OtherEndClosed`. This error allows us to implement a function like `read_all(fd) -> Vec<u8>` to read all the data from the pipe at once. Many programming languages provide similar functions in their standard libraries, and you can easily implement this functionality yourself in the CKB-VM.
```

- The last line closes the file descriptor.

```rust
#[no_mangle]
pub unsafe extern "C" fn main() -> u64 {
    let fds = inherited_fds();
    let out = ARGS.join("");
    write(fds[0], out.as_bytes());
    close(fds[0]);
    return 0;
}
```

So, how to use this library? Follow the following code:

- The first line defines the arguments to be passed to the child Script.
- The second line creates a pipe.
- The third line calls the `spawn()` function, passing in the arguments and the writable file descriptor.
- The fourth line reads all the data from the readable file descriptor.
- The last line uses the `wait()` function to wait until the child Script exits.

```rust
#[no_mangle]
pub unsafe extern "C" fn main() -> u64 {
    let argv = ["Hello", "World!"];
    let fds = pipe();
    let pid = spawn_cell(code_hash, hash_type, &argv, &[fds[1]]);
    let data = read_all(fds[0]);
    assert_eq!(&data, b"Hello World!");
    wait(pid);
    return 0;
}
```

## Enhanced Efficiency With Spawn’s Pipes

Since `spawn()` operates within its own memory space, independent of the parent process, it can mitigate certain security risks, particularly in scenarios involving sensitive data.

> While some other blockchain virtual machines, such as EVM, also offer cross-contract call functions, these calls tend to be quite expensive, especially when invoking child contracts repeatedly. This is primarily because it requires continuously creating and destroying a virtual machine to complete each call, pictured as below:

Reviewer: ```suggestion
While some other blockchain virtual machines, such as the EVM, also offer cross-contract call functions, these calls tend to be quite expensive, especially when invoking child contracts repeatedly. This is primarily because each call requires continuously creating and destroying a virtual machine, as illustrated below:
```

                      |                  | Pipe |       | Pipe |       |
              Create  |                  |  IO  |       |  IO  |       | Destroy
                      v                  |      |       |      |       |
                   +--+-------+          v      |       v      |       |
                   | Script B |----------+------+-------+------+-------+
                   +----------+
```

## Typical Use Cases

Spawn has several potential applications, and we believe it's particularly advantageous in the following scenarios:

- **Public Library**: Spawn enables the development of common libraries, allowing different Scripts to reuse shared code. This approach simplifies Script development and substantially reduces deployment costs. For instance, we could publish diverse cryptographic algorithms on CKB as public libraries.
- **Large Script Management**: When your Script's binary size exceeds 500K, it surpasses CKB's maximum transaction size limit for Script deployment. In such cases, you can leverage Spawn to divide and deploy your Script logic separately, effectively circumventing this limitation.

> #@ Conclusion

Reviewer: ```suggestion
## Conclusion
```
